### PR TITLE
Some more fixes for compressed meshes

### LINKED ISSUE #83240: Mesh compression breaks normal
### Godot version

v4.2.beta.custom_build [ee118e7ff]

### System information

Godot v4.2.beta (ee118e7ff) - Windows 10.0.22621 - Vulkan (Forward+) - dedicated AMD Radeon RX 6700 XT (Advanced Micro Devices, Inc.; 27.20.21034.37) - AMD Ryzen 7 5700X 8-Core Processor (16 Threads)

### Issue description

When mesh compression is enabled (default), the normals of a mesh becomes corrupted. Most likely introduced by #81138 (needs verification). May be related to #82890. Seemed to start appearing after I overrode light materials.

With compression:
image
image

Without compression:
image
image

### Steps to reproduce

(See MRP)
1. Check normal buffer.
2. Go to the import settings of `dm3_open_sky.obj` and disable `Force Disable Compression`.
3. Check normal buffer again.

### Minimal reproduction project

[UV2 Compression.zip](https://github.com/godotengine/godot/files/12890164/UV2.Compression.zip)

### LINKED ISSUE #83526: Shadow meshes give error `Can't use compression flag 'ARRAY_FLAG_COMPRESS_ATTRIBUTES'` and null `ARRAY_VERTEX`
### Godot version

4.2.beta1 a574c0296b

### System information

Godot v4.2.beta (a574c0296) - Windows 10.0.19045 - Vulkan (Forward+) - dedicated NVIDIA GeForce RTX 3090 (NVIDIA; 31.0.15.3713) - Intel(R) Core(TM) i9-9900K CPU @ 3.60GHz (16 Threads)

### Issue description

Calling `Mesh.surface_get_arrays()` on a shadow mesh appears to print errors such as 
```
  Can't use compression flag 'ARRAY_FLAG_COMPRESS_ATTRIBUTES' while using normals without tangent array.
  scene\resources\mesh.cpp:1795 - Condition "err != OK" is true.
  scene\resources\mesh.cpp:1936 - Index p_idx = 0 is out of bounds (surfaces.size() = 0).
  scene\resources\mesh.cpp:1916 - Index p_idx = 0 is out of bounds (surfaces.size() = 0).
```
and returns arrays with the `ARRAY_VERTEX` array being null.

My code does:
```gd
	func adjust_mesh_scale(mesh: ArrayMesh, is_shadow: bool = false):
		push_warning("Processing mesh " + str(mesh) + " is_shadow=" + str(is_shadow))
		for surf_idx in range(surf_count):
			var prim: int = mesh.surface_get_primitive_type(surf_idx)
			var fmt_compress_flags: int = mesh.surface_get_format(surf_idx)
			var arr: Array = mesh.surface_get_arrays(surf_idx)

			push_warning("Surface " + str(surf_idx) + " : Arrays len is " + str(len(arr)) + " (should be " + str(ArrayMesh.ARRAY_INDEX + 1) + ")")
			push_warning("Surface " + str(surf_idx) + " : ARRAY_VERTEX (" + str(ArrayMesh.ARRAY_VERTEX) + ") should be type " + str(typeof(PackedVector3Array())) + " : " + str(typeof(arr[0])))
			push_warning("Surface " + str(surf_idx) + " : len(arr[ARRAY_VERTEX]) is.....")
			push_warning(len(arr[0])) # Crashes on surface 1?????

		...

		if not is_shadow and mesh.shadow_mesh != mesh and mesh.shadow_mesh != null:
			adjust_mesh_scale(mesh.shadow_mesh, true)
```

Here is the output on this function when processing both the conventional mesh and the shadow mesh:
```
  core\variant\variant_utility.cpp:1033 - Processing mesh <ArrayMesh#-9223369631605918152> is_shadow=false
  core\variant\variant_utility.cpp:1033 - Surface 0 : Arrays len is 13 (should be 13)
  core\variant\variant_utility.cpp:1033 - Surface 0 : ARRAY_VERTEX (0) should be type 36 : 36
  core\variant\variant_utility.cpp:1033 - Surface 0 : len(arr[ARRAY_VERTEX]) is.....
  core\variant\variant_utility.cpp:1033 - 8280
  Can't use compression flag 'ARRAY_FLAG_COMPRESS_ATTRIBUTES' while using normals without tangent array.
  scene\resources\mesh.cpp:1795 - Condition "err != OK" is true.
  scene\resources\mesh.cpp:1936 - Index p_idx = 0 is out of bounds (surfaces.size() = 0).
  scene\resources\mesh.cpp:1916 - Index p_idx = 0 is out of bounds (surfaces.size() = 0).
  core\variant\variant_utility.cpp:1033 - Processing mesh <ArrayMesh#-9223369631589142756> is_shadow=true
  core\variant\variant_utility.cpp:1033 - Surface 0 : Arrays len is 13 (should be 13)
  core\variant\variant_utility.cpp:1033 - Surface 0 : ARRAY_VERTEX (0) should be type 36 : 0
  core\variant\variant_utility.cpp:1033 - Surface 0 : len(arr[ARRAY_VERTEX]) is.....
  res://test_post_import.gd:271 - Error calling GDScript utility function '<unknown function>': Value of type 'Nil' can't provide a length.
```

### Steps to reproduce

1. Open the attached project.
2. Clear the console
3. Reimport `Godot.glb`

### Minimal reproduction project

[mesh_post_import.zip](https://github.com/godotengine/godot/files/12962578/mesh_post_import.zip)

### PR DESCRIPTION
**Pull request by clayjohn**
This cleans up a few more cases of uint32_t->uint64_t

Importantly this fixes an edge case in the axis-angle compression by using the pre-existing Basis methods instead

Fixes: https://github.com/godotengine/godot/issues/83240
Fixes: https://github.com/godotengine/godot/issues/83526

This might help https://github.com/godotengine/godot/issues/83667, but I wasn't able to reproduce, so I don't know for sure. 

For posterity, most of the changes here are just cleaning up uint32_t->uint64_t and other QoL changes I caught while debugging. The biggest change is the change to ``_get_axis_angle()``. Instead of using the small helper function, I switched to using ``Basis::_get_axis_angle()`` which has the proper check for if the matrix is symmetrical. The issue we were running into was that some meshes contain bogus tangents (https://github.com/godotengine/godot/issues/83240 is full of them). This can happen with meshes that don't have UVs set up for normalmapping and have the "generate_tangents" flag checked by default. 

Sfter spending hours trying to understand the math behind the compression and find any source for the original code. I stumbled on https://en.wikipedia.org/wiki/Rotation_matrix#Conversion_from_rotation_matrix_to_axis%E2%80%93angle which explains everything perfectly and derives the steps exactly how we have them laid out in the code. It contains this important warning:

> This does not work if R is symmetric. ... In this case, it is necessary to diagonalize R and find the eigenvector corresponding to an eigenvalue of 1.

In searching for ``is_symmetric()``, I found ``diagonalize()`` which led me to ``Basis::_get_axis_angle()`` which contained all the proper checks and cleaning to properly handle bogus tangents. Accordingly, I switched over to using Basis entirely and everything works nicely.

### CODE REVIEW

File: servers/rendering_server.cpp [hunk 1]
```diff
@@ -1399,7 +1397,7 @@ Array RenderingServer::_get_array_from_surface(uint64_t p_format, Vector<uint8_t
 								tangentsw[j * 4 + 3] = tan.w;
 							}
 							ret[RS::ARRAY_NORMAL] = normals;
-							ret[RS::ARRAY_FORMAT_TANGENT] = tangents;
+							ret[RS::ARRAY_TANGENT] = tangents;
```
Comment (clayjohn) [inline servers/rendering_server.cpp:1400]: This alone fixes https://github.com/godotengine/godot/issues/83526. What a silly mistake

**Review by akien-mga — APPROVED** [review]

**Review by AThousandShips — APPROVED** [review]

**Comment by fire** [comment]
I looked at these compressed mesh fixes, but I wasn't able to see anything obviously wrong.

File: scene/resources/mesh.cpp [hunk 2]
```diff
@@ -2008,17 +2015,19 @@ void ArrayMesh::regen_normal_maps() {
 		return;
 	}
 	Vector<Ref<SurfaceTool>> surfs;
+	Vector<uint64_t> formats;
 	for (int i = 0; i < get_surface_count(); i++) {
 		Ref<SurfaceTool> st = memnew(SurfaceTool);
 		st->create_from(Ref<ArrayMesh>(this), i);
 		surfs.push_back(st);
+		formats.push_back(Ref<ArrayMesh>(this)->surface_get_format(i));
```
Comment (bitsawer) [inline scene/resources/mesh.cpp:2023]:
```suggestion
		formats.push_back(surface_get_format(i));
```

**Comment by clayjohn** [comment]
Updated based on comments!

File: scene/resources/mesh.cpp [hunk 3]
```diff
@@ -1639,6 +1639,13 @@ void ArrayMesh::_set_surfaces(const Array &p_surfaces) {
 		uint64_t surface_version = surface.format & (ARRAY_FLAG_FORMAT_VERSION_MASK << ARRAY_FLAG_FORMAT_VERSION_SHIFT);
 		if (surface_version != ARRAY_FLAG_FORMAT_CURRENT_VERSION) {
 			RS::_fix_surface_compatibility(surface);
+			surface_version = surface.format & (RS::ARRAY_FLAG_FORMAT_VERSION_MASK << RS::ARRAY_FLAG_FORMAT_VERSION_SHIFT);
+			ERR_FAIL_COND_MSG(surface_version != RS::ARRAY_FLAG_FORMAT_CURRENT_VERSION,
+					"Surface version provided (" +
+							itos((surface_version >> RS::ARRAY_FLAG_FORMAT_VERSION_SHIFT) & RS::ARRAY_FLAG_FORMAT_VERSION_MASK) +
+							") does not match current version (" +
+							itos((RS::ARRAY_FLAG_FORMAT_CURRENT_VERSION >> RS::ARRAY_FLAG_FORMAT_VERSION_SHIFT) & RS::ARRAY_FLAG_FORMAT_VERSION_MASK) +
+							").");
```
Comment (akien-mga) [inline scene/resources/mesh.cpp:1648]: Could be slightly more readable (IMO) with `vformat`:
```suggestion
			ERR_FAIL_COND_MSG(surface_version != RS::ARRAY_FLAG_FORMAT_CURRENT_VERSION,
					vformat("Surface version provided (%d) does not match current version (%d).",
							(surface_version >> RS::ARRAY_FLAG_FORMAT_VERSION_SHIFT) & RS::ARRAY_FLAG_FORMAT_VERSION_MASK),
							(RS::ARRAY_FLAG_FORMAT_CURRENT_VERSION >> RS::ARRAY_FLAG_FORMAT_VERSION_SHIFT) & RS::ARRAY_FLAG_FORMAT_VERSION_MASK)));
```

If you want to change it, you can wait for #83613 to be merged as there will be a merge conflict anyway.

**Comment by clayjohn** [comment]
Fixed up the error message and rebased!

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**